### PR TITLE
AV-174357 add unittest for public ip feature

### PR DESCRIPTION
**Pull request by arihantg**

### CODE REVIEW

**Comment by arihantg** [comment]
build amko

File: gslb/test/integration/third_party_vips/gslb_hostrule_test.go [hunk 1]
```diff
@@ -875,3 +875,55 @@ func TestGSLBHostRuleHmRefToHealthMonitorTemplate(t *testing.T) {
...
+		}, 5*time.Second, 1*time.Second).Should(gomega.Equal(true))
+	}
+	hostName := routeObj.Spec.Host
+	oldObj := addGSLBHostRule(t, gslbHRName, gslbutils.AVISystem, hostName, hmRefs, nil, nil, &ttl,
+		ingestion.GslbHostRuleAccepted, "")
+	gslbhostrulepip := []gslbalphav1.PublicIPElem{
+		{Cluster: K8sContext, IP: "10.20.30.40"},
+		{Cluster: OshiftContext, IP: "10.20.40.50"}}
+	newObj := getGSLBHostRule(t, oldObj.Name, oldObj.Namespace)
+	newObj.Spec.PublicIP = gslbhostrulepip
```
Comment (swathinsankaran) [inline gslb/test/integration/third_party_vips/gslb_hostrule_test.go:912]: This seems to be wrong. We have to assign the IP to `thirdPartyMembers`.

File: gslb/test/integration/third_party_vips/gslb_hostrule_test.go [hunk 2]
```diff
@@ -875,3 +875,55 @@ func TestGSLBHostRuleHmRefToHealthMonitorTemplate(t *testing.T) {
...
+	//update hostrule to add public IP
+	updateGSLBHostRule(t, newObj, ingestion.GslbHostRuleAccepted, "")
+	verifyMembers(gslbhostrulepip)
+	newGslbhostrulepip := []gslbalphav1.PublicIPElem{
+		{Cluster: K8sContext, IP: "10.20.30.50"},
+		{Cluster: OshiftContext, IP: "10.20.60.50"}}
+	newObj = getGSLBHostRule(t, oldObj.Name, oldObj.Namespace)
+	newObj.Spec.PublicIP = newGslbhostrulepip
```
Comment (swathinsankaran) [inline gslb/test/integration/third_party_vips/gslb_hostrule_test.go:920]: `ThirdPartyMembers[i].PublicIP` must be used here instead.

File: helm/amko/crds/gslbhostrule_def.yaml [hunk 3]
```diff
@@ -129,6 +129,9 @@ spec:
                     ip:
                       description: "Public IP of the corresponding site"
                       type: string
+                  required:
```
Comment (swathinsankaran) [inline helm/amko/crds/gslbhostrule_def.yaml:132]: we must take care of this in the `pkg/apis/v1alpha1/types.go`. The `omitempty` is not required if the field is required.

**Comment by arihantg** [comment]
build amko

File: gslb/nodes/node_utils.go [hunk 4]
```diff
@@ -140,6 +140,8 @@ func setGSLBPropertiesForGS(gsFqdn string, gsGraph *AviGSObjectGraph, newObj boo
 			gsGraph.MemberObjs[idx].Priority = getK8sMemberPriority(priorityMap, member.Cluster, member.Namespace)
 			if pIP, ok := publicIPMap[member.Cluster]; ok {
 				gsGraph.MemberObjs[idx].PublicIP = pIP
+			} else {
+				gsGraph.MemberObjs[idx].PublicIP = ""
 			}
```
[the remark this answers is not in the document]
Comment (arihantg) [reply gslb/nodes/node_utils.go]: moved to a new function

File: gslb/test/ingestion/gslbhostrule_test.go [hunk 5]
```diff
@@ -229,3 +229,35 @@ func TestGSLBHostRuleInvalidDownResponse(t *testing.T) {
 	g.Expect(err.Error()).Should(gomega.Equal("Fallback IP INVALID is not valid"))
 	t.Logf("Verified GSLBHostRule")
 }
+
+func TestGSLBHostRuleValidPublicIP(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	buildAndAddTestGSLBObject(t)
+	// GSLBHostRule with down response of type GSLB_SERVICE_DOWN_RESPONSE_FALLBACK_IP and empty fallbackIP.
```
[the remark this answers is not in the document]
Comment (arihantg) [reply gslb/test/ingestion/gslbhostrule_test.go]: comment modified

File: gslb/test/ingestion/gslbhostrule_test.go [hunk 6]
```diff
@@ -229,3 +229,35 @@ func TestGSLBHostRuleInvalidDownResponse(t *testing.T) {
...
+	err := gslbingestion.ValidateGSLBHostRule(gslbhrObj, false)
+	t.Logf("Verifying GSLBHostRule")
+	g.Expect(err).To(gomega.BeNil())
+	t.Logf("Verified GSLBHostRule")
+}
+
+func TestGSLBHostRuleInvalidPublicIP(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+	buildAndAddTestGSLBObject(t)
+	// GSLBHostRule with down response of type GSLB_SERVICE_DOWN_RESPONSE_FALLBACK_IP and empty fallbackIP.
```
[the remark this answers is not in the document]
Comment (arihantg) [reply gslb/test/ingestion/gslbhostrule_test.go]: comment modified

File: gslb/test/ingestion/gslbhostrule_test.go [hunk 7]
```diff
@@ -229,3 +229,35 @@ func TestGSLBHostRuleInvalidDownResponse(t *testing.T) {
...
+	gslbhrObj := getTestGSLBHRObject(gslbhrTestObjName, gslbhrTestNamespace, gslbhrTestFqdn)
+	gslbhrObj.Spec.PublicIP = []gslbalphav1.PublicIPElem{{Cluster: "k8", IP: "10.23.23.45"}, {Cluster: "cluster2", IP: "10.23.23.46"}}
+
+	err := gslbingestion.ValidateGSLBHostRule(gslbhrObj, false)
+	t.Logf("Verifying GSLBHostRule")
+	g.Expect(err).NotTo(gomega.BeNil())
+	g.Expect(err.Error()).Should(gomega.Equal("cluster k8 in Public IP  not present in GSLBConfig"))
+	gslbhrObj.Spec.PublicIP = []gslbalphav1.PublicIPElem{{Cluster: "cluster1", IP: "10.23.23"}, {Cluster: "cluster2", IP: "10.23.23.46"}}
```
[the remark this answers is not in the document]
Comment (arihantg) [reply gslb/test/ingestion/gslbhostrule_test.go:264]: Have added valid and invalid ipv6 test

File: gslb/test/integration/third_party_vips/gslb_hostrule_test.go (hunk 1 above)
Comment (arihantg) [reply gslb/test/integration/third_party_vips/gslb_hostrule_test.go:912]: added another Ut to test third party

File: helm/amko/crds/gslbhostrule_def.yaml (hunk 3 above)
Comment (arihantg) [reply helm/amko/crds/gslbhostrule_def.yaml:132]: removed

**Review by swathinsankaran — APPROVED** [review]
LGTM

**Review by akshayhavile — APPROVED** [review]

**Pull request merged**